charmhelpers/contrib/network/ip.py

```diff
@@ -519,14 +531,23 @@ def get_hostname(address, fqdn=True):
                 apt_install("python3-dnspython", fatal=True)
             import dns.reversename
 
+        result = None
+        exc_value = None
         rev = dns.reversename.from_address(address)
-        result = ns_query(rev)
-
-        if not result:
-            try:
-                result = socket.gethostbyaddr(address)[0]
-            except Exception:
-                return None
+        try:
+            result = ns_query(rev)
+        except Exception:
+            # If reverse DNS lookup failed, try nsswitch (may check /etc/hosts)
+            # If that fails too, raise this exception/traceback instead.
+            (exc_type, exc_value, exc_traceback) = sys.exc_info()
+        finally:
+            if not result:  # Hit exception (non-NXDOMAIN) or no lookup/answer.
+                try:
+                    result = socket.gethostbyaddr(address)[0]
+                except Exception:
+                    if exc_value:
+                        reraise(exc_type, exc_value, exc_traceback)
+                    return None
```

Comment:
Again does this do exactly the same thing?

Reply:
Likewise.

Reply:
Ah, indeed, see above.